File: lib/svg/elements/round/circle.py
from ..svgElementClass import SvgElement
from ..utils.attributeUpdater import update
from ..utils import randint


class Circle(SvgElement):
    isEmpty = True
    name = "circle"

    def __init__(
        self, radius: int, center: tuple[int, int] = (0, 0), inner=None, outer=None
    ):
        self.r = radius
        self.x = center[0]
        self.y = center[1]

        attributes = [
            ("r", f"{self.r}"),
            ("cx", f"{self.x}"),
            ("cy", f"{self.y}"),
        ]

        super().__init__(Circle.name, attributes, inner, outer, isEmpty=True)

    def changeCenter(self, x, y):
        update("cx", f"{x}", self.attributes)
        update("cy", f"{y}", self.attributes)

        self.x = int(x)
        self.y = int(y)

    @staticmethod
    def generate(height, width):
        r = randint(min(height, width))
        x, y = randint(width), randint(height)

        attributes = [
            ("r", f"{r}"),
            ("cx", f"{x}"),
            ("cy", f"{y}"),
        ]

        return SvgElement.generate(Circle.name, attributes, Circle.isEmpty)

File: lib/svg/elements/round/circle.cc
#include "circle.hh"

#include "../utils/mathHelpers.hh"

Circle::Circle(size_t radius, const std::tuple<size_t, size_t>& center,
               std::optional<Coloring> inner, std::optional<Outline> outer)
    : Round("circle",
            { { "r", std::to_string(radius) },
              { "x", std::to_string(std::get<0>(center)) },
              { "y", std::to_string(std::get<1>(center)) } },
            radius, center, inner, outer)
{}

Circle Circle::generate(size_t height, size_t width)
{
    size_t r = randint(MIN(height, width));
    size_t x = randint(width);
    size_t y = randint(height);

    return Circle(r, { x, y }, Coloring::generate(), Outline::generate());
}